### PR TITLE
avro-extensions -- feature to specify avro reader schema inline

### PR DESCRIPTION
**Pull request by himanshug**
in the task json for all events.

See https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/druid-development/tg2q3VHN5rQ/5kUuiTouEQAJ

### CODE REVIEW

**Comment by skoppar** [comment]
This is something we need in our set up. Thanks Himanshu. Which release is this expected to be a part of ?

File: extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ */
+public class SimpleAvroBytesDecoder implements AvroBytesDecoder
+{
+  private static final Logger logger = new Logger(SimpleAvroBytesDecoder.class);
+
+  private final Schema schemaObj;
+
+  private Map<String, Object> schema;
```
Comment (drcrallen) [inline extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java]: final?

File: extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      @JsonProperty("schema") Map<String, Object> schema
+  ) throws Exception
+  {
+    Preconditions.checkArgument(schema != null, "schema must be provided");
+
+    this.schema = schema;
+    String schemaStr = mapper.writeValueAsString(schema);;
+
+    logger.info("Schema string [%s]", schemaStr);
+    schemaObj = new Schema.Parser().parse(schemaStr);
```
Comment (drcrallen) [inline extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java]: how do schema errors manifest themselves here? Will they show up in the hadoop Reader?

File: extensions-core/avro-extensions/src/test/java/io/druid/data/input/avro/SimpleAvroBytesDecoderTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import com.fasterxml.jackson.databind.InjectableValues;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.druid.jackson.DefaultObjectMapper;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ */
+public class SimpleAvroBytesDecoderTest
```
Comment (drcrallen) [inline extensions-core/avro-extensions/src/test/java/io/druid/data/input/avro/SimpleAvroBytesDecoderTest.java]: is it possible to build some Avro data for parsing as part of this test?

File: extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+{
+  private static final Logger logger = new Logger(SimpleAvroBytesDecoder.class);
+
+  private final Schema schemaObj;
+
+  private Map<String, Object> schema;
+
+  @JsonCreator
+  public SimpleAvroBytesDecoder(
+      @JacksonInject ObjectMapper mapper,
```
Comment (drcrallen) [inline extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java]: Suggest `@Json` annotation if you're wanting the json mapper explicitly.

**Comment by drcrallen** [comment]
This should be reasonable to be in 0.9.2, and if there's a community need even more so

File: extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java (hunk 1 above)
Comment (himanshug) [inline extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java]: added

File: extensions-core/avro-extensions/src/test/java/io/druid/data/input/avro/SimpleAvroBytesDecoderTest.java (hunk 3 above)
Comment (himanshug) [inline extensions-core/avro-extensions/src/test/java/io/druid/data/input/avro/SimpleAvroBytesDecoderTest.java]: added

File: extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java (hunk 4 above)
Comment (himanshug) [inline extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java]: added

File: extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java (hunk 2 above)
Comment (himanshug) [inline extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/SimpleAvroBytesDecoder.java]: for that we can rely on avro lib only , if there is an error in the schema parsing, then this should throw exceptions.
otoh if the schema is ok but datum given in the parse is not deserializable then we should get an exception there.

**Comment by himanshug** [comment]
@skoppar do you have single schema for all the events in the kafka topic or multiple ? is it possible for you to test this patch?

are you already trying to use avro extension with schema_repo and facing some error?

**Comment by nishantmonu51** [comment]
LGTM, 👍

**Comment by himanshug** [comment]
updated the docs.

**Comment by skoppar** [comment]
Hi Himanshu, 
I havent tried Schema repo and was about to extend the AvroBytesDecoder myself and thats when I saw your comment on another google thread. https://groups.google.com/forum/#!searchin/druid-user/avro$20druid%7Csort:relevance/druid-user/72NKnm6eBRY/VCgyaTKEAwAJ

**Comment by skoppar** [comment]
Hi Himanshu,

I am not able to get the avro stream to work with the new jar. I have
packaged your changes which produced
"druid-avro-extensions-0.9.2-SNAPSHOT.jar".
I moved the original jar into a subdirectory so that there are no conflicts
and copied the snapshot jar in current directory (imply-1.3.0/dist/druid/
extensions/druid-avro-extensions). Also changed imply-1.3.0/conf-
quickstart/druid/_common/common.runtime.properties to include avro library:

druid.extensions.loadList=["druid-avro-extensions", "druid-histogram",
"druid-datasketches", "druid-kafka-indexing-service"]


I've modified kafka.json in imply-1.3.0/conf-quickstart/tranquility to the
one attached.

Upon starting imply:
bin/supervise -c conf/supervise/quickstart.conf

I do not see any topics being picked in the tranquility logs
[imply-1.3.0/var/sv/tranquility-kafka.log]:
2016-08-08 17:27:56,913 [main] INFO  k.c.ZookeeperConsumerConnector -
[cust1-tranquility_skoppar-mac15.com-1470677276325-37ef097f], Creating
topic event watcher for topics ((?!))
2016-08-08 17:27:56,929 [main] INFO  k.c.ZookeeperConsumerConnector -
[cust1-tranquility_skoppar-mac15.com-1470677276325-37ef097f], Topics to
consume = List()


Doesn't it mean that no topics have been registered?

Only relevant entries in the log are:
2016-08-08 17:27:56,151 [main] INFO  c.m.t.kafka.writer.WriterController -
Ready: [topicPattern] -> dataSource mappings:
2016-08-08 17:27:56,152 [main] INFO  c.m.t.kafka.writer.WriterController -
  [(?!)] -> AvroSource (priority: 1)

I am new to Druid, so please pardon my ignorance. Is the
publicAvrokafka.json file right? At this point I havent added any business
functionality. The derived fields are coming from an ETL process outside of
Druid. I just want to be able to ingest avro data through kafka and that
doesnt seem to work. To be sure its not a kafka issue, I have tried using
kafka-console-consumer and I can receive events there.

regards
Sunita

**Comment by himanshug** [comment]
@skoppar i haven't really used kafka with tranquility and information provided does not look like avro parsing related failures.
i don't see any kafka topic etc input information in your spec , not sure how you provide it to tranquility... @gianm can you spot something?

**Comment by skoppar** [comment]
Thanks for quick response Himanshu.
From what I understand,
"properties": {
  "topicPattern": "raw-cust1"

}

Provides the topic which is per datasource level

and below properties provide overall cluster connection. Adding
group_id to data source specs might

be more advisable, but I haven't tried that yet. I am using all default ports.

"kafka.group.id":"cust1-tranquility",

"zookeeper.connect":"localhost",

"kafka.zookeeper.connect":"localhost"


regards

Sunita

**Comment by skoppar** [comment]
Another fact is, Zookeeper is able to recognize the consumer group and the
group is visible upon executing zookeeper shell. Also from tranquility
side, zookeeper connection seems to be present. The Kafka broker connection
is what seems to be missing and the topic list is empty as mentioned in the
previous comment. Appreciate any help  @gianm <https://github.com/gianm>

**Comment by nishantmonu51** [comment]
@skoppar seems like you have specified the properties at wrong place in json spec file, 
the properties needs to be defined outside the spec file as done here - 
https://github.com/druid-io/tranquility/blob/master/distribution/src/universal/conf/kafka.json.example

**Comment by skoppar** [comment]
Thanks Nishant for bringing it to my notice.
Yea I see there is some discrepancy in the spec now. Will check that and
update this space

**Comment by skoppar** [comment]
Hi Himanshu,

Looks like there is a dependency on io.druid.druid-api-0.9.1.jar based on the errors I see. imply-1.3.0/dist/tranquility/lib has a huge list of jars. I hope I replace fewer jars. Are there any documented dependencies? Attaching the exception I get now. 
[WithPR.txt](https://github.com/druid-io/druid/files/409649/WithPR.txt)

**Comment by gianm** [comment]
@skoppar what is the tranquility config you are using now?

**Comment by skoppar** [comment]
@gianm attaching the current tranquility-kafka json file.
[publicAvroKafka.txt](https://github.com/druid-io/druid/files/409671/publicAvroKafka.txt)

**Comment by himanshug** [comment]
@skoppar @gianm given that exception came from StringInputRowParser , I believe avro_stream parser did not get recognized and that probably means avro-extensions module wasn't loaded.  did you see any avro-extension loading failures or a message regarding loading of avro-extension in the logs?

**Comment by skoppar** [comment]
@himanshug doesnt look like loading the jar is the issue. I could be wrong though. Attaching the indexing-log in imply-1.3.0/var/druid/indexing-logs which shows the druid-avro-extensions-0.9.2-SNAPSHOT.jar being added. Attaching the indexing log. Please let me know if I should be looking at a wrong place. 
[index_realtime_metrics-kafka.txt](https://github.com/druid-io/druid/files/409703/index_realtime_metrics-kafka.txt)

**Comment by himanshug** [comment]
@skoppar one of the problems in your json is that "parseSpec" is inside "avroBytesDecoder" while it should be a level up inside "parser" .

**Comment by skoppar** [comment]
oops ! let me relook at the complete json

**Comment by skoppar** [comment]
Its working now. 2 more changes I had to do was:
1. Specify the format in parseSpec
2. Add the path to druid extensions manually in quickStart.conf 

With that done to the attached conf, I am able to process records with schema inline.

**Comment by himanshug** [comment]
@skoppar glad to know that it worked out for you and thanks for testing the patch.

**Comment by fjy** [comment]
👍

**Comment by skoppar** [comment]
Well not completely. Trying to figure out why segments are not getting
created. But the patch did work well. This must be something else I am
missing. I have 800 messages received and unparseableCount=800. I tried
changing timestamp to be current, but still get the same results.
Appreciate any inputs you have on this, however, it does not seem to be
related to the changes in the PR.

On Aug 10, 2016, at 10:47 AM, Himanshu <notifications@github.com> wrote:

@skoppar <https://github.com/skoppar> glad to know that it worked out for
you and thanks for testing the patch.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/druid-io/druid/pull/3249#issuecomment-238946405>, or mute
the thread

**Comment by skoppar** [comment]
Well not completely. Trying to figure out why segments are not getting
created. But the patch did work well in terms of pulling the schema and
getting right count of events. This must be something else I am missing. I
have 800 messages received and unparseableCount=800. I tried changing
timestamp to be current, but still get the same results. Appreciate any
inputs you have on this, however, it probably is not related to the changes
in the PR.


On Aug 10, 2016, at 10:47 AM, Himanshu <notifications@github.com> wrote:

@skoppar <https://github.com/skoppar> glad to know that it worked out for
you and thanks for testing the patch.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/druid-io/druid/pull/3249#issuecomment-238946405>, or mute
the thread

**Comment by himanshug** [comment]
@skoppar enable debug logging and that would show you more info.

**Comment by skoppar** [comment]
Figured it out with the help of a kafka consumer class. The console consumer showed data just fine, so couldn't spot the error. The rootcause is, I was sending the avro message using StringEncoder. Using DefaultEncoder on producer side fixed the issue. Thanks @himanshug and @gianm. Awaiting 9.2 release :) 👍

**Pull request merged**